garage/tf/models/base.py

```diff
@@ -0,0 +1,241 @@
+"""This file contains the abstraction class for models."""
+import abc
+import collections
+import warnings
+
+import tensorflow as tf
+
+from garage.tf.models.auto_pickable import PickleCall
+from garage.tf.models import AutoPickable
+
+
+class BaseModel(abc.ABC):
+    """Interface-only abstract class for models."""
+    @property
+    def init_spec(self):
+        """Positional and named arguments (args and kwargs) which, when passed
+        to `__init__`, will construct this model"""
+        pass
+
+    def build(self, *inputs):
+        """Output of model with the given input placeholder(s)."""
+        pass
+
+    @property
+    def name(self):
+        """A unique name for this Model."""
+        pass
+
+    @property
+    def input(self):
+        """Tensor input of the Model."""
+        pass
+
+    @property
+    def output(self):
+        """Tensor output of the Model."""
+        pass
+
+    @property
+    def inputs(self):
+        """Tensor inputs of the Model."""
+        pass
+
+    @property
+    def outputs(self):
+        """Tensor outputs of the Model."""
+        pass
+
+    @property
+    def weights(self):
+        """Weights of the Model."""
+        pass
+
+    @weights.setter
+    def weights(self, weights):
+        """Set weights of the Model."""
+        pass
+
+
+class Model(BaseModel, metaclass=abc.ABCMeta):
+    """Abstract class for models with sane defaults."""
+    def __init__(self):
+        self._name = type(self).__name__  # name default to class
+        self._inputs = None
+        self._outputs = None
+        self._default_weights = None
+        self._built = False
+
+    def build(self, *inputs):
+        """Output of model with the given input placeholder(s)."""
+
+        # Generally, you should not be rebuilding models
```

Comment:
So can we rebuild a model with customized inputs?

Reply:
If you want to rebuild a model then you want to make a new subgraph in TF for that model. In which case you should copy the model, not rebuild the same model.

Forbidding rebuilding keeps a 1:1 Python object:TF subgraph mapping.